### PR TITLE
Meeting page

### PR DESCRIPTION
**Pull request by Codball**
Referencing issue #58 
Added functionality for past meetings. Shows paginated meetings containing the date, speakers and their talk title. Mostly compatible with mobile. The only thing remaining is how to implement showing a link to the meetings page on the homepage.

Requires an interesting workaround for paginating a struct with preloaded associations. See comment in meetings_controller.ex.

chrome_2019-01-28_16-49-24

### CODE REVIEW

File: lib/cbus_elixir/app/app.ex [hunk 1]
```diff
@@ -20,9 +19,11 @@ defmodule CbusElixir.App do
 
   """
   def list_speakers do
-    query = from s in Speaker,
+    query = from(s in Speaker,
             join: m in assoc(s, :meeting),
             order_by: [desc: m.date]
```
Comment (Kartstig) [inline lib/cbus_elixir/app/app.ex]: I know this isn't in scope of this PR, but it'd be really cool if you can move the `Repo.preload([:meeting])` on L29 to inside this query

File: lib/cbus_elixir/app/meetings.ex [hunk 2]
```diff
@@ -7,6 +7,51 @@ defmodule CbusElixir.App.Meetings do
...
+        join: s in assoc(m, :speakers),
+        where: m.date < ^DateTime.utc_now(),
+        order_by: [desc: m.date],
+        preload: [speakers: s]
+      )
+
+      Pagination.paginate(query, page, per_page)
+  end
+
+
```
Comment (Kartstig) [inline lib/cbus_elixir/app/meetings.ex]: Would love to add typespecs for all public functions in here. Looks like:
```elixir
@spec attendees_for_meeting(integer()) :: [Ecto.schema.t()]
```

File: lib/cbus_elixir/app/meetings.ex [hunk 3]
```diff
@@ -7,6 +7,51 @@ defmodule CbusElixir.App.Meetings do
...
+    query = 
+      from(a in Attendee,
+        where: a.meeting_id == ^id,
+        select: a.name
+      )
+
+    query
+    |> Repo.all()
+  end
+
```
Comment (Kartstig) [inline lib/cbus_elixir/app/meetings.ex:42]: Add typespec

File: lib/cbus_elixir/app/meetings.ex [hunk 4]
```diff
@@ -7,6 +7,51 @@ defmodule CbusElixir.App.Meetings do
...
+  def attendees_for_meetings(id) do
```
Comment (Kartstig) [inline lib/cbus_elixir/app/meetings.ex]: This function name is confusing. It seems to be returning Meetings, but the function name implies that the return type would be `Attendee.t()`

File: lib/cbus_elixir/pagination.ex [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  def paginate(query, page, per_page) do
+    count = total_entries(query) |> Repo.one()
+    
+
+    paged_results = paged_query(query, page, per_page) |> Repo.all()
+    {results, has_next} = split_results(per_page, paged_results)
+
+    has_next = has_next
```
Comment (Kartstig) [inline lib/cbus_elixir/pagination.ex]: This line is meaningless since `has_next` gets assigned on L14

File: lib/cbus_elixir/pagination.ex [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  defp paged_query(query, page, per_page) do
+
+    offset = per_page * (page - 1)
+
+    query
+  |> limit(^(per_page + 1))
+  |> offset(^offset)
+  end
+
+  defp split_results(per_page, paged_results) when length(paged_results) <= per_page do
```
Comment (Kartstig) [inline lib/cbus_elixir/pagination.ex:37]: Wondering if there's a better name for this. It does indeed split the results, but part of this also seems to calculate `has_next`

File: lib/cbus_elixir/pagination.ex [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  def paginate(query, page, per_page) when is_binary(page) do
+    paginate(query, String.to_integer(page), per_page)
+  end
+
+  def paginate(query, page, per_page) do
+    count = total_entries(query) |> Repo.one()
+    
+
+    paged_results = paged_query(query, page, per_page) |> Repo.all()
+    {results, has_next} = split_results(per_page, paged_results)
```
Comment (Kartstig) [inline lib/cbus_elixir/pagination.ex:13]: At first glance, I was confused about the calculation of `has_next`, but it does seem like preloading (i+1) and checking to see if it's there is pretty slick

File: lib/cbus_elixir/pagination.ex [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      {h, t} -> {h, true}
+    end
+  end
+
+  defp total_entries(query) do
+    total_entries =
+      query
+      |> exclude(:preload)
+      |> exclude(:select)
+      |> subquery
```
Comment (Kartstig) [inline lib/cbus_elixir/pagination.ex]: 👮 : Please add `()`

File: lib/cbus_elixir_web/controllers/meeting_controller.ex [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  alias CbusElixir.App.Meetings
+  alias CbusElixir.Repo
+
+  def index(conn, params) do
+    page = params["page"] || 1
+    per_page = params["per_page"] || 5
+    
+    meetings = Meetings.meetings_for_page(page, per_page)
+
+    IO.inspect(meetings)
```
Comment (Kartstig) [inline lib/cbus_elixir_web/controllers/meeting_controller.ex]: This should be removed

File: lib/cbus_elixir_web/controllers/meeting_controller.ex [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+    render(conn, "index.html", meetings: meetings)
+  end
+
+  def show(conn, %{"id" => id}) do
+    attendees = Meetings.attendees_for_meeting(id)
+    IO.inspect(attendees)
```
Comment (Kartstig) [inline lib/cbus_elixir_web/controllers/meeting_controller.ex]: Remove

File: lib/cbus_elixir/pagination.ex [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+      last: if(rem(count, per_page) == 0, do: div(count, per_page), else: div(count, per_page) + 1),
+      count: count,
+      results: results
+    }
+  end
+
+  
+
+  defp paged_query(query, page, per_page) do
+
```
Comment (Kartstig) [inline lib/cbus_elixir/pagination.ex]: Remove extra line

**Review by Kartstig — CHANGES_REQUESTED** [review]
Looks good overall. I have some suggestions

**Comment by croz1007** [comment]
@Kartstig @Codball I can't seem to locate this branch when I do a `git fetch` so I cannot even pull it locally to try and get it running. Hmmm, I'll check again later. Also, can you please provide screenshots and/or gifs of the screens and a link to the issue this is related to as part of the description for this PR?

**Comment by Kartstig** [comment]
@croz1007 you'll have to add the remote for @codball 's fork. We don't give push access to public users

File: lib/cbus_elixir/app/meetings.ex (hunk 4 above)
Comment (Codball) [reply lib/cbus_elixir/app/meetings.ex]: Yeah that's an extra function I was trying to implement in the beginning, I ended up going with the simpler, more efficient query above. Will remove.

File: lib/cbus_elixir/app/app.ex (hunk 1 above)
Comment (Codball) [reply lib/cbus_elixir/app/app.ex]: Which query do you think it should be moved to?

**Comment by croz1007** [comment]
Still trying to get it running. Looks like I may have a lot of npm updates to make. We'll see.

**Comment by croz1007** [comment]
Got it running locally and reviewing now.

**Comment by croz1007** [comment]
Something is very sideways on this right now. I tried to create a new speaker for the 2/5/19 meeting and when I did it added that speaker to all of the remaining meetings. This is not happening on `master`. Looking into it a bit more now. Just wanted to call it out in case you know where to change before I locate it. Thanks

File: priv/repo/seeds.exs [hunk 12]
```diff
@@ -24,3 +25,131 @@ Enum.each(0..100, fn x ->
   |> Meeting.changeset(%{date: meeting_date})
   |> Repo.insert!()
 end)
+
```
Comment (croz1007) [inline priv/repo/seeds.exs]: We do not want to have speakers seeded in the app. The reason being in that we have the see file setting up the meeting dates for us. We would run this in the event we have an issue with our server and needed to setup the dates again. Seeding speakers would put a lot of bogus data in the DB at the time we setup our dates again if we needed to do that.

**Review by ieatkimchi — CHANGES_REQUESTED** [review]
we are using the seeds to setup meetings dates so we don't want to create speakers in the seeds. Also we need to add some test coverage here on the new code to verify if it works.

File: priv/repo/seeds.exs (hunk 12 above)
Comment (croz1007) [reply priv/repo/seeds.exs]: If you are needing Speakers to test this with you can either A) create new speakers using the UI at `speakers/new` (you would just need to comment out the basic auth in the speaker_controller to get past the authentication, or B) Add some speakers to the DB via the command line.

**Comment by Codball** [comment]
At the moment pagination breaks when you add multiple speakers. The query meetings_for_page/2 in App.Meetings.ex returns the correct information, but when passed to paged_query/3 in Pagination.ex the results returned are mixed up in an odd way, I can't recognize a pattern. Is it counting a speaker as part of the limit? I'm thinking I could grab the speakers in a different way, perhaps in another query or preloading them after the query is paginated.

File: lib/cbus_elixir/app/meetings.ex [hunk 13]
```diff
@@ -5,12 +5,51 @@ defmodule CbusElixir.App.Meetings do
...
   alias CbusElixir.App.Meeting
+  alias CbusElixir.Pagination
+  alias CbusElixir.App.Speaker
+  alias CbusElixir.App.Attendee
+
+  @doc """
+  Fetch a paged result for meetings
+  """
+  
+  @spec meetings_for_page_query() :: Ecto.Query
```
Comment (Kartstig) [inline lib/cbus_elixir/app/meetings.ex]: should be `Ecto.Query.t()` or `Query.t()`

File: lib/cbus_elixir/app/meetings.ex [hunk 14]
```diff
@@ -5,12 +5,51 @@ defmodule CbusElixir.App.Meetings do
...
+        preload: [speakers: s]
+      )
+
+  end
+
+  @doc """
+  Returns the attendees for the given meeting
+  """
+
+  @spec attendees_for_meeting(integer()) :: map()
```
Comment (Kartstig) [inline lib/cbus_elixir/app/meetings.ex]: This should actually be returning `list(map())` or more specifically `list(Meeting.t())`

File: lib/cbus_elixir/app/meetings.ex [hunk 15]
```diff
@@ -29,3 +70,7 @@ defmodule CbusElixir.App.Meetings do
     |> Repo.preload([:speakers, :attendees])
   end
 end
+
+
+
+
```
Comment (Kartstig) [inline lib/cbus_elixir/app/meetings.ex]: Extra whitespace

**Comment by Kartstig** [comment]
Some notes:

For the Next and Previous buttons, you can try to use FontAwesome. Here's a great one for Next:
https://fontawesome.com/icons/angle-right?style=solid

Then you just have to do the following:
```html
<i class="fas fa-angle-right"></i>
```

**Comment by Kartstig** [comment]
When you add tests, make sure you make changes to prevent seeds from running multiple times in your database: https://github.com/columbus-elixir/marketing/pull/64

Then you can do length based validations. Otherwise, you'll have tons of meetings shoved into your test db after running it many times.

**Comment by Kartstig** [comment]
@Codball Your tests should pass locally now if you pull from master

File: lib/cbus_elixir/app/meetings.ex [hunk 16]
```diff
@@ -5,12 +5,48 @@ defmodule CbusElixir.App.Meetings do
...
   alias CbusElixir.Repo
   alias CbusElixir.App.Meeting
+  alias CbusElixir.Pagination
+  alias CbusElixir.App.Speaker
+  alias CbusElixir.App.Attendee
+
+  @doc """
+  Fetch past meetings
+  """
+
```
Comment (Kartstig) [inline lib/cbus_elixir/app/meetings.ex]: Extra line between @doc and @spec

File: lib/cbus_elixir/app/meetings.ex [hunk 17]
```diff
@@ -5,12 +5,48 @@ defmodule CbusElixir.App.Meetings do
...
+      where: m.date < ^DateTime.utc_now(),
+      order_by: [desc: m.date],
+      distinct: m.date
+    )
+  end
+
+  @doc """
+  Returns the attendees for the given meeting
+  """
+
```
Comment (Kartstig) [inline lib/cbus_elixir/app/meetings.ex]: Extra space

File: lib/cbus_elixir/app/meetings.ex [hunk 18]
```diff
@@ -5,12 +5,48 @@ defmodule CbusElixir.App.Meetings do
...
+  Fetch past meetings
+  """
+
+  @spec meetings_for_page_query() :: Ecto.Query.t()
+  def meetings_for_page_query() do
+    from(m in Meeting,
+      left_join: s in assoc(m, :speakers),
+      where: m.date < ^DateTime.utc_now(),
+      order_by: [desc: m.date],
+      distinct: m.date
```
Comment (Kartstig) [inline lib/cbus_elixir/app/meetings.ex]: Oh I see. So you're able to get past the limit result issue by calling distinct?

File: lib/cbus_elixir/app/meetings.ex [hunk 19]
```diff
@@ -5,12 +5,48 @@ defmodule CbusElixir.App.Meetings do
...
+      )
+
+    query
+    |> Repo.all()
+  end
 
   @doc """
   Returns the next `count` upcoming meetings
   """
+
```
Comment (Kartstig) [inline lib/cbus_elixir/app/meetings.ex]: Was this a result of the formatter?

File: test/cbus_elixir/app/pagination_test.exs [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+      days = Enum.find(0..6, fn d -> Timex.shift(date, days: d) |> Date.day_of_week() == 2 end)
+      meeting_date = Timex.shift(date, days: days)
+
+      %Meeting{}
+      |> Meeting.changeset(%{date: meeting_date})
+      |> Repo.insert!()
+    end)
+  end
+
+  describe "paginate/3" do
```
Comment (Kartstig) [inline test/cbus_elixir/app/pagination_test.exs:21]: :100:

File: lib/cbus_elixir_web/templates/meeting/show.html.eex [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+    </th>
+    <tr>
+      <td><%= attendee %></td>
+    </tr>
+  </table>
+
+<% end %>
+
+
+
```
Comment (Kartstig) [inline lib/cbus_elixir_web/templates/meeting/show.html.eex]: Lots of extra whitespace here

File: lib/cbus_elixir/app/meetings.ex (hunk 19 above)
Comment (Codball) [reply lib/cbus_elixir/app/meetings.ex]: No, but the formatter doesn't correct it, either.

**Review by Kartstig — DISMISSED** [review]
Tested this locally and it looks great. Thanks for working through all the review comments!
:100: :1st_place_medal: :+1:

**Comment by Kartstig** [comment]
Do we want to put a link to this page from `index`?

**Comment by croz1007** [comment]
@Kartstig @Codball 
I think that a link on the index page would be great. I don't believe that it is a top nav type of link, but rather a link down below the "This Month's Speakers" section with a heading of "Past meet ups" or something like that. Looks like you have a heading there in your screenshots.

Just a thought about that link: Could either be just a link to that `/meetings` page, or we could list out the dates of the last 3 as links to those meetings and a generic link to see all past meetings. I guess that we are not doing a `show` page for a specific meeting at this point so we would not be able to link to them individually right now. 

Also, Nathan, I think the updated screenshots look really great.

**Review by croz1007 — APPROVED** [review]
Looks really great!

**Review by ieatkimchi — APPROVED** [review]

**Comment by ieatkimchi** [comment]
looks good, fyi there is an ecto lib for pagination. https://github.com/drewolson/scrivener_ecto

**Pull request merged**